File: python/test/test_camera.py
import math
import unittest

from rtc.camera import Camera
from rtc.color import Color
from rtc.transform import Transform, ViewTransform
from rtc.tuples import *
from rtc.utils import req
from rtc.world import DefaultWorld


class TestCamera(unittest.TestCase):
    def test_camera_constructor(self):
        hsize = 160
        vsize = 120
        fov = math.pi / 2
        c = Camera(hsize, vsize, fov)

        self.assertEqual(c.hsize, 160)
        self.assertEqual(c.vsize, 120)
        self.assertEqual(c.fov, math.pi / 2)
        self.assertEqual(c.transform, Transform())

    def test_camera_horizontal_pixel_size(self):
        c = Camera(200, 125, math.pi / 2)
        self.assertTrue(req(c.pixel_size, 0.01))

    def test_camera_vertical_pixel_size(self):
        c = Camera(125, 200, math.pi / 2)
        self.assertTrue(req(c.pixel_size, 0.01))

    def test_ray_through_center(self):
        c = Camera(201, 101, math.pi / 2)
        r = c.ray_for_pixel(100, 50)
        self.assertTrue(tuple4_eq(r.origin, Point(0, 0, 0)))
        self.assertTrue(tuple4_eq(r.direction, Vector(0, 0, -1)))

    def test_ray_through_corner(self):
        c = Camera(201, 101, math.pi / 2)
        r = c.ray_for_pixel(0, 0)
        self.assertTrue(tuple4_eq(r.origin, Point(0, 0, 0)))
        self.assertTrue(tuple4_eq(r.direction, Vector(0.66519, 0.33259, -0.66851)))

    def test_ray_when_transformed(self):
        c = Camera(201, 101, math.pi / 2)
        c.transform = Transform().translation(0, -2, 5).rotation_y(math.pi / 4)
        r = c.ray_for_pixel(100, 50)
        self.assertTrue(tuple4_eq(r.direction, Vector(math.sqrt(2) / 2, 0, -math.sqrt(2) / 2)))
        self.assertTrue(tuple4_eq(r.origin, Point(0, 2, -5)))

    def test_camera_render(self):
        w = DefaultWorld()
        c = Camera(11, 11, math.pi / 2)
        at = Point(0, 0, -5)
        to = Point(0, 0, 0)
        up = Vector(0, 1, 0)
        c.transform = ViewTransform(at, to, up)
        image = c.render(w)
        self.assertEqual(image.read(5, 5), Color(0.38066, 0.47583, 0.2855))

File: tests/test_camera.c

#include "tests.h"

void test_camera(void)
{
  TESTS();

  TEST {
      // Default camera
      camera c = {0};
      camera_init(&c, 160, 120, PI_2);

      assert(c.hsize == 160);
      assert(c.vsize == 120);
      assert(req(c.fov, PI_2));
      assert(m4_eq(c.transform, IDENTITY));
  }

  TEST {
      // Pixel size for horizontal canvas
      camera c = {0};
      camera_init(&c, 200, 125, PI_2);

      assert(req(c.pixel_size, 0.01));
  }

  TEST {
      // Pixel size for vertical canvas
      camera c = {0};
      camera_init(&c, 125, 200, PI_2);

      assert(req(c.pixel_size, 0.01));
  }

  TEST {
      // Ray through center of canvas
      camera c = {0};
      camera_init(&c, 201, 101, PI_2);

      ray r = {0};

      camera_ray_for_pixel(&c, 100, 50, &r);

      assert(v4_eq(r.origin, point(0, 0, 0)));
      assert(v4_eq(r.direction, vector(0, 0, -1)));
  }

  TEST {
      // Ray through corner of canvas
      camera c = {0};
      camera_init(&c, 201, 101, PI_2);

      ray r = {0};

      camera_ray_for_pixel(&c, 0, 0, &r);

      assert(v4_eq(r.origin, point(0, 0, 0)));
      assert(v4_eq(r.direction, vector(0.66519, 0.33259, -0.66851)));
  }

  TEST {
      // Ray through canvas when camera is transformed
      camera c = {0};
      camera_init(&c, 201, 101, PI_2);

      {
        m4 R = {0};
        rotation_y(PI_4, R);

        m4 T = {0};
        translation(0, -2, 5, T);

        m4 Z = {0};
        m4_mul(R, T, Z);

        camera_set_transform(&c, Z);
      }

      ray r = {0};

      camera_ray_for_pixel(&c, 100, 50, &r);

      assert(v4_eq(r.origin, point(0, 2, -5)));
      assert(v4_eq(r.direction, vector(ROOT_2_2, 0, -ROOT_2_2)));
  }

  TEST {
      world w = {0};
      world_init(&w);

      camera v = {0};
      camera_init(&v, 11, 11, PI_2);

      v4 from = point_init(0, 0, -5);
      v4 to = point_init(0, 0, 0);
      v4 up = vector_init(0, 1, 0);

      m4 T = {0};
      view_transform(from, to, up, T);

      camera_set_transform(&v, T);

      canvas *c = camera_render(&v, &w, NULL);
      assert(c != NULL);

      v3 *px = canvas_at(c, 5, 5);
      assert(px != NULL);

      assert(v3_eq(*px, color(0.38066, 0.47583, 0.2855)));

      canvas_free(c);
  }
}

